Optimization, error handlers

Alexcitten opened the pull request:
**Possible error handlers were added, optimization was done, and a bit of beauty was added.**

The initialization of Arweave was moved outside of the `deployContract` function, since there is no need to initialize it several times.
Then created a separate `deployContract` function to handle the deployment process. Renamed some variables for clarity. And also added error handling for the deploy method

I used ES6 to import individual functions, and simplified operators to check input and error handling.
Created a separate functionMap object to map input functions to the corresponding function implementations.
Useful to have error handling when the input function is not in the function map.
Renamed the change-price function to `changePrice` to follow the standard naming conventions in the camel case, I understand that such "cutting" of code is not necessary, but why not.

Used destructive assignment to get the input and caller properties from the action object, and the default value syntax to set the target variable to an undefined value if it is not present in the input object looks nice.

I also changed the `ContractAssert` condition so that it is not executed when the target is false, moved the calculation of the balance variable after the `ContractAssert` condition to avoid unnecessary calculations when the target is not present.
Simplified the balance calculation with the `||` operator so that the default value is 0 when the target is not present in the balance object.

I used destructive assignment to get the input and caller properties from the action object, and changed the second ContractAssert condition so that it doesn't execute if target is false.
A good solution is to move `balances[caller] -= qty` before `balances[target] += qty` to avoid double-counting when caller and target are the same.
Added check to remove caller from balances when caller no longer owns tokens.

Alexcitten committed:
deploy.js improvements 🛠

• Moved the Arweave initialization outside of the deployContract function since it's not necessary to initialize it multiple times.
• Created a separate function deployContract to handle the deployment process.
• Used template literals for string interpolation.
• Renamed some variables to be more descriptive.
• Added error handling for the deploy method.

Alexcitten committed:
contract.js improvements 🛠

• Used ES6 import syntax to import the individual functions.
• Simplified the if statement for input validation and error handling.
• Created a separate object functionMap to map input functions to their corresponding function implementations.
• Added error handling for the case where the input function is not in the function map.
• Renamed the change-price function to changePrice to follow standard camel case naming conventions.

Alexcitten committed:
balance.js improvements 🛠

• Used destructuring assignment to get the input and caller properties from the action object.
• Used default value syntax to set the target variable to undefined if it is not present in the input object.
• Changed the ContractAssert condition to fail when target is falsy.
• Moved the calculation of the balance variable to after the ContractAssert condition to avoid unnecessary computation when target is not present.
• Simplified the balance calculation using the || operator to default to 0 when target is not present in the balances object.

Alexcitten committed:
transfer.js improvements 🛠

• Used destructuring assignment to get the input and caller properties from the action object.
• Changed the second ContractAssert condition to fail when target is falsy.
• Changed the third ContractAssert condition to use a period at the end of the sentence and added a space before the last period for consistency.
• Moved the balances[caller] -= qty line before the balances[target] += qty line to avoid double-counting when caller and target are the same.
• Added a check to delete caller from balances when caller no longer owns any tokens.

angrymouse reviewed (changes requested):
Can you also add contract code address to output? It's useful property

angrymouse commented on src/contract.js:
```diff
@@ -1,19 +1,32 @@
+import transfer from "./functions/transfer.js";
```
Will imports work with esbuild's IIFE setup?
It can potentially break execution

angrymouse commented on src/functions/balance.js:
```diff
@@ -1,19 +1,21 @@
 module.exports = async function (state, action) {
```
if you use ESM import, use ESM export as well, or it doesn't work

Alexcitten committed:
deploy.js improvements 🛠

The address of the contract code was added to the output
Removed `return`, because code don't need to interact with the deployed contract from another function or module at this point.

Alexcitten committed:
contract.js improvements 🛠

Replacing the ES6 import statements with require() calls.

Alexcitten committed:
balance.js improvements 🛠

`module.exports = async function (state, action) { ` => `export async function handle(state, action) {`

Alexcitten committed:
transfer.js improvements 🛠

module.exports = async function (state, action) {  => export async function handle(state, action) {

Alexcitten commented:
Yes, you are correct. I should have thought about that, and checked it out.
Kind of fixed it.

angrymouse commented on deploy.js:
```diff
@@ -31,7 +31,9 @@ async function deployContract() {
 
-  return deployResult;
+  // Displaying the deployed contract code
+  const contractCode = await arweave.transactions.getData(srcTxId, { decode: true, string: true });
```
You don't have to re-download code... What I meant is showing Transaction ID of code that you've uploaded to arweave

Alexcitten committed:
deploy.js improvements 🛠

Alexcitten committed:
Added comments

Added comments to explain what each part of the code does

Alexcitten committed:
finalize-buy.js improvements 🛠

• Extracted owner, caller, and input from the action to make the code more readable.
• Improved the error message for the second assertion by clarifying that only the reserver can finalize the buy.
• Reordered the assertions to make them easier to read.
• Added semicolons to the end of the statements for consistency.

Alexcitten committed:
change-price.js improvements 🛠

• Using strict equality (===) instead of loose equality (==) to compare state.owner and action.caller
• Storing the price value from action.input in a separate variable to avoid using it multiple times in the assertion and improving readability
• Using Number.isFinite(price) to check that the price is a finite number
• Checking that the price is greater than 0, since a price of 1 or lower may not make sense in the context of the contract.

Alexcitten committed:
list.js improvements 🛠

• Used strict equality operator === instead of loose equality == for better type checking
• Used semicolons for statement termination
• Used single quotes consistently for string literals
• Made the error message consistent with other error messages by starting with a lowercase letter and ending with a period
• Clarified the error message for the ownership check by changing "Should own NFT" to "Caller should own NFT" for better readability
• Moved the state modifications to the end of the function for better readability

Alexcitten committed:
reserve-buying-zone.js improvements 🛠

• Used destructuring to extract the properties target, quantity, owner, and id from SmartWeave.transaction.
• Changed the comparison operator from == to === for better type checking.
• Used strict equality operator (===) instead of loose equality operator (==) for better type checking.
• Used template literals to combine string and variable values.
• Added semicolons to the end of each statement for consistency.
• Used object literal shorthand syntax when creating the return object.

Alexcitten committed:
unlist.js improvements 🛠

• Removed unnecessary semicolons.
• Added comments to explain what the function does.
• Renamed the variables to be more descriptive.
• Moved the state.forSale check to the top, to make the function exit early if the NFT is already not for sale.
• Reset the reservation variables to null, since the NFT is no longer for sale.
• Return the modified state object wrapped in an object.

Alexcitten commented:
I hope I've done everything right now.
A description of each improvement can be found in the comet description
